docs/proposals/config_data.md

```diff
@@ -0,0 +1,298 @@
+* Create a flexible configuration model for ``Kubernetes``.
+
+## Motivation
+
+Currently command-line flags and environment variables are used for configurations in ``Kubernetes``
+which is complicated and error-prone. This approach is not suitable for a distributed computing environment.
+Specifically the problems with the current approach are:
+
+* Complicated deployment orchestration
+* Hard to synchronize across replicated components.
+* Hard to change dynamically as needed by continuous services.
+
+## Use Cases
+
+1. Be able to load API resources as environment variables into a component.
+2. Be able to load API resources as mounted volumes into a component.
+3. Be able to update mounted volumes after a change to the API resource.
+
+## Solution
+
+A new ``ConfigData`` API resource is proposed to address these problems.
+
+The ``ConfigData`` architecture will be very similar to that of ``Secrets``.
+
+``ConfigData`` can either be mounted as a volume or can be injected as environment
+variables to a container through the use of the downward API.
+
+### Dynamic Configuration
+
+Any long-running system has mutating specification over time. In order to facilitate this functionality,
+``ConfigData`` will be versioned and updates will automatically made available to the container.
+
+``resourceVersion`` (Found in ``ObjectMetadata``) of the ``ConfigData`` object will be updated by the API server every time the object is modified.  After an update, modifications will be made visible to the consumer container. If the consumer uses the ``Data`` pairs only for initialization or during starting process, A rolling-update might be necessary to update the components.
+
+It is then the consumer component's responsibility to make use of the updated data
+once it is made visible (or perform a rolling-update on consumers of that object. This is especially true if ``ConfigData`` is injected as environment variables. Dynamic configuration with environment variables is out of scope).
+
+### Advantages
+
+* Easy distribution through the API server.
+* Provides configuration data in a consumer-agnostic manner.
+* Persistent configuration information through API versioning.
+* Ability to use ``Kubernetes`` authentication and security policies with configuration objects.
+* By leveraging the power of ``/watch`` API or watching the volume, container processes can implement responsiveness to configuration changes.
+
+## API Resource
+
+A new resource for ``ConfigData`` will be added to the ``Extentions`` API Group:
+
+```go
+	type ConfigData struct {
+		TypeMeta   `json:",inline"`
+		ObjectMeta `json:"metadata,omitempty"`
+
+		Data map[string]string `json:"data,omitempty"`
+	}
+
+	type ConfigDataList struct {
+		TypeMeta `json:",inline"`
+		ListMeta `json:"metadata,omitempty"`
+
+		Items []ConfigData `json:"items"`
+	}
+```
+
+A new API resource selector (``ConfigDataSelector``) will be added to select objects and fields from those objects
```

Comment:
I think there should be a new subsection here detailing the specific API changes needed for consuming configuration in envs.